### ISSUE TITLE
All datasets don't have an option to get validation set split.

### ISSUE THREAD

**Issue by gagb** [pytorch/vision#168]
Why do only some datasets have the options to get a validation set split?
For example, LSUN dataset has this option, MNIST and CIFAR10 don't.

Or is there an option and I missed it?

**Comment by rishab96** [comment]
I'm not sure about MNIST, but LSUN has a validation set in its official dataset, while CIFAR does not. You can easily create a validation set in CIFAR using this code:

`class ChunkSampler(sampler.Sampler):
    """Samples elements sequentially from some offset. 
    Arguments:
        num_samples: # of desired datapoints
        start: offset where we should start selecting from
    """
    def __init__(self, num_samples, start = 0):
        self.num_samples = num_samples
        self.start = start

    def __iter__(self):
        return iter(range(self.start, self.start + self.num_samples))

    def __len__(self):
        return self.num_samples

NUM_TRAIN = 49000
NUM_VAL = 1000

cifar10_train = dset.CIFAR10('./cs231n/datasets', train=True, download=True,
                           transform=T.ToTensor())
loader_train = DataLoader(cifar10_train, batch_size=64, sampler=ChunkSampler(NUM_TRAIN, 0))

cifar10_val = dset.CIFAR10('./cs231n/datasets', train=True, download=True,
                           transform=T.ToTensor())
loader_val = DataLoader(cifar10_val, batch_size=64, sampler=ChunkSampler(NUM_VAL, NUM_TRAIN))`

**Comment by fmassa** [comment]
As @rishab96 has mentioned, not all datasets have an official validation split.

**Comment by alyato** [comment]
@fmassa 
im trying the pytorch firstly. 
i used to use the keras and the dataset has 3 parts , train,valid,test.
but when i check the https://github.com/pytorch/examples/blob/master/mnist/main.py, it has train function and test function .
I cannot find the valid_dataset,only the train_loader and test_loader
As what @gagb say, doesn't the valid_dataset need to exist.
Im using to classify task.
So my dataset don't need the valid_dataset and only have train_dataset and test_dataset?
is it right?
thanks

**Comment by vfdev-5** [comment]
@alyato the example :  https://github.com/pytorch/examples/blob/master/mnist/main.py you are citing it is just an example.
Yes, you still need train/valid/test
and you can split train dataset into two with a random splitting, for example like this:
```python
from torch.utils.data.dataset import Subset

# Random splitting
train_indices = ...
val_indices = ...

trainval = MNIST('../data', train=True, download=True, transform=...)
train_ds = Subset(trainval, train_indices)
val_ds = Subset(trainval, val_indices)
```
and go on with training on train dataset, validate on `val_ds` and finally test of testing dataset

HTH

**Comment by alyato** [comment]
@vfdev-5 
thanks.
yes. I do split the dataset as what you say.
Then, how can i use this train_dataset,valid_dataset,test_dataset.
i check the [https://github.com/pytorch/examples/blob/master/mnist/main.py](url),it present an example include the `train` `function` and `test` `function`.But it doesn't present the `valid_dataset`.
Is there one example show me that the train_dataset using for training, the valid_dataset using for adjust the parameters and the test_datset using for testing?
yes, i don't how to implement the `valid_function` code.
thanks again.

**Comment by vfdev-5** [comment]
@alyato sorry do not really get your question, but if you would like to have different transformation for valid dataset from train dataset, you can do like this:
```
# Random splitting
train_indices = ...
val_indices = ...

train_ds = MNIST('../data', train=True, download=True, transform=train_transform)
val_ds = MNIST('../data', train=True, download=True, transform=val_transform)

from torch.utils.data.sampler import SubsetRandomSampler
from torch.utils.data import DataLoader

train_sampler = SubsetRandomSampler(train_indices)
val_sampler = SubsetRandomSampler(val_indices)
train_loader = DataLoader(train_ds, sampler=train_sampler)
val_loader = DataLoader(train_ds, sampler=val_sampler)
```
HTH
